### PR TITLE
feat(memory_diff): back change ledger with git instead of SQLite

### PR DESCRIPTION
**Pull request by senamakel**
## Summary

- Replace the `memory_diff` SQLite snapshot store (`diff.db`) with a **git repository** at `<workspace>/memory_diff/repo` as the change ledger.
- Map the diff domain onto git primitives: **snapshot → commit** (`Snapshot.id` is the commit SHA), **checkpoint → annotated tag** `ckpt_<uuid>`, **read marker → ref** `refs/openhuman/read/<source_id>`, **diff → git tree diff** scoped to the source subtree.
- Keep the entire public API unchanged — all 10 RPC handlers, the `memory_diff` agent tool, the subconscious tick, and `auto_snapshot_after_sync` keep their signatures; only the storage/diff engine swapped.
- Add vendored `git2`/libgit2 (no system-git or network transports) so it works on end-user desktop machines.
- `cleanup` now prunes old checkpoint tags only; git retains full history (delta-compressed).

## Problem

The previous `memory_diff` module reimplemented in SQLite a lot of what git does natively: per-item content hashing, a manual added/removed/modified loop, `similar`-based text diffs capped at 2 KB, a 64 KB per-item content bound, named checkpoints via join tables, read markers, and 30-day/100-snapshot retention. This is exactly git's wheelhouse (content-addressed blobs, tree diffs with real patch output, tags, refs, delta-compressed history).

## Solution

`mem_tree_chunks` stays authoritative; the git repo is a *derived* ledger. Each `take_snapshot` materialises a source's current items as one blob per item under `<source_id>/` and commits, carrying every other source forward from the parent commit so HEAD always reflects the whole world. Item identity is the (encoded) filename, `content_hash` is the blob oid, and text diffs come from real git patches. Snapshot metadata that has no natural git home (kind/label/trigger/count/ms-timestamp) rides in commit trailers.

Design decisions:
- **Single-writer lock**: all mutations serialise through a process-global `Mutex` because libgit2's HEAD/parent resolution is read-modify-write (SQLite WAL previously handled this).
- **Flat, encoded item filenames** (`encode_item_id`, prefixed `i_`) keep item ids with slashes/colons/reserved names (`.`/`..`) git-safe and reversible; content change → `Modified`, id rename → `Removed`+`Added` (same id-keyed semantics as before).
- **Starts clean** (no `diff.db` migration, per design discussion) — the first post-upgrade diff shows existing items as added.

## Submission Checklist

> If a section does not apply to this change, mark the item as `N/A` with a one-line reason. Do not delete items.

- [x] Tests added or updated (happy path + at least one failure / edge case) per [Testing Strategy](../gitbooks/developing/testing-strategy.md#failure-path-requirement)
- [x] **Diff coverage ≥ 80%** — 33 unit tests (git_store + ops) plus the JSON-RPC e2e lifecycle test cover the rewritten lines including failure/edge cases (cross-source rejection, pathspec-prefix isolation, missing-marker fallback, empty source subtree). CI `coverage-gate` enforces the threshold.
- [x] N/A: Coverage matrix — behaviour-preserving backend swap; no feature rows added/removed/renamed (existing `memory_diff` coverage is unchanged).
- [x] N/A: No matrix feature IDs changed by this PR.
- [x] No new external network dependencies introduced — `git2` is built with `default-features = false, features = ["vendored-libgit2"]` (no SSH/HTTPS transports); all git operations are on a local repo.
- [x] N/A: Manual smoke checklist — internal change-tracking storage; no release-cut/user-facing surface change.
- [x] N/A: No tracking issue for this change.

## Impact

- **Platform**: desktop/CLI (Rust core). Adds a C compile (vendored libgit2) to the core build — the build already compiles whisper/llama/CEF, so no new toolchain requirement; cold build of the dep is ~2 min.
- **Migration**: ledger starts clean; legacy `diff.db` is simply no longer read (forward-only). First diff after upgrade reports existing items as added.
- **Behaviour**: `cleanup`'s `deleted_snapshots` count now reflects pruned checkpoint *tags*; snapshot commits are retained.
- **Security**: no PII/secret logging added; verbose `[memory_diff]`/`[memory_diff::git]` tracing on init/commit/diff/tag.

## Related

- Closes: N/A
- Follow-up PR(s)/TODOs: N/A

---

### Linear Issue

- Key: N/A
- URL: N/A

### Commit & Branch

- Branch: feat/memory-diff-git
- Commit SHA: dd346200da5496d1efa8325a0366fa80896ec15a

### Validation Run

- [x] `pnpm --filter openhuman-app format:check`
- [x] `pnpm typecheck`
- [x] Focused tests: `cargo test --lib memory_diff` (33 passed); `test-rust-with-mock.sh --test json_rpc_e2e json_rpc_memory_diff_snapshot_diff_and_read_marker_lifecycle` (1 passed)
- [x] Rust fmt/check (if changed): `cargo check --lib` + `cargo clippy` clean
- [x] Tauri fmt/check (if changed): `pnpm rust:check` (pre-push) passed

### Validation Blocked

- `command:` N/A
- `error:` N/A
- `impact:` N/A

### Behavior Changes

- Intended behavior change: storage/diff engine swapped SQLite → git; `cleanup` semantics narrowed to checkpoint tags.
- User-visible effect: none in the RPC/tool contract; diffs/checkpoints/read-markers behave identically.

### Parity Contract

- Legacy behavior preserved: all RPC handlers, agent tool, subconscious tick, and `auto_snapshot_after_sync` keep signatures and semantics (verified by ported unit tests + unchanged e2e).
- Guard/fallback/dispatch parity checks: cross-source diff still rejected; missing/stale read marker treated as unread; pathspec isolation prevents prefixed-source leakage.

### Duplicate / Superseded PR Handling

- Duplicate PR(s): N/A
- Canonical PR: N/A
- Resolution (closed/superseded/updated): N/A

### CODE REVIEW

File: src/openhuman/memory_diff/git_store.rs [hunk 1]
```diff
@@ -0,0 +1,868 @@
...
+
+        let repo = match Repository::open(&repo_path) {
+            Ok(repo) => repo,
+            Err(_) => {
+                tracing::debug!(
+                    path = %repo_path.display(),
+                    "[memory_diff::git] initialising diff ledger repository"
+                );
+                Repository::init(&repo_path)
+                    .with_context(|| format!("init memory_diff repo: {}", repo_path.display()))?
```
[the remark this answers is not in the document]
Comment (senamakel) [reply src/openhuman/memory_diff/git_store.rs:83]: **Dismissed (intentional design decision):** Starting the git ledger clean — with no read/fallback of the legacy `diff.db` — is the deliberate, product-owner-approved choice for this PR (see the PR description: "Starts clean (no `diff.db` migration, per design discussion)" and the Migration note under Impact).

The `memory_diff` ledger is a *derived* view of `mem_tree_chunks` (the authoritative store), not a primary datastore, so the only post-upgrade effect is cosmetic and self-healing: `diff_since_read` replays current items as unread once, then the marker advances; a fresh ledger does not reference old `diff.db` checkpoint IDs, and the subconscious tick re-establishes a baseline on its next run. A seeded-migration path was explicitly considered and declined in favour of a forward-only cutover.
